### PR TITLE
Fix isotonic regression gradients for arbitrary input ranks

### LINKED ISSUE #127234: tf.nn.isotonic_regression returns incorrect gradient for a pooled block
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.21.0, also reproduced on tf 2.22.0-dev20260904

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Mobile device

_No response_

### Python version

Python 3.13.5

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

`tf.nn.isotonic_regression` returns an incorrect first derivative for a stable pooled block.

The reproducer evaluates decreasing isotonic regression on `[4, t, 2]` with `t = 0.7`. For `0 < t < 2`, the last two entries form a stable pooled block with value `(t + 2) / 2`. Therefore, the selected output `y[1]` has derivative `1/2` throughout an open neighborhood.

The forward value is computed correctly as `1.35`, but reverse-mode autodiff returns `1.0` for the derivative with respect to `t`. The expected derivative is `0.5`.

Expected behavior?

The first derivative of the selected `tf.nn.isotonic_regression` output at `t = 0.7` should be close to `0.5`, not `1.0`.

### Standalone code to reproduce the issue

```shell
import os
os.environ["CUDA_VISIBLE_DEVICES"] = ""
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["TF_NUM_INTRAOP_THREADS"] = "1"
os.environ["TF_NUM_INTEROP_THREADS"] = "1"

import tensorflow as tf

def target(t):
    x = tf.stack([
        tf.constant(4, dtype=t.dtype),
        t,
        tf.constant(2, dtype=t.dtype),
    ])
    y, segments = tf.nn.isotonic_regression(
        inputs=x,
        decreasing=True,
        axis=0,
    )
    return y[1]

x = tf.constant(0.7, dtype=tf.float64)

with tf.GradientTape() as tape:
    tape.watch(x)
    y = target(x)

actual = tape.gradient(y, x)
expected = 0.5

print("forward:", y.numpy())
print("actual:", actual.numpy())
print("expected:", expected)

if abs(float(actual) - expected) > 1e-9:
    print("BUG REPRODUCED: tf.nn.isotonic_regression returns incorrect gradient for a pooled block")
else:
    print("not reproduced")
```

### Relevant log output

```shell
forward: 1.35
actual: 1.0
expected: 0.5
BUG REPRODUCED: tf.nn.isotonic_regression returns incorrect gradient for a pooled block
```

### PR DESCRIPTION
**Pull request by twelfthlabor**
Fixes #127234. (The 1-D case was fixed by #127550; this PR generalizes the fix to arbitrary ranks, empty and dynamic shapes, and is rebased on top of it.)

The isotonic gradient assumed a two-dimensional input. For `[4., 0.7, 2.]`, differentiating the middle output returned `1.0` with an extra dimension instead of the scalar `0.5`. Higher-rank inputs, empty inputs and unknown shapes also failed.

Average gradients using row-offset segment IDs, then restore the input shape. This keeps rows independent and removes the Python loop over a statically known batch size. Tests cover ranks 1–3, both ordering directions, non-last axes, float32/float64, eager/graph execution, dynamic shapes, and empty inputs. The test BUILD target includes the new GradientTape dependency.

Rebase notes (head `f8481721`, base master `72c0c0c9`):
- Master's merged 1-D special case in `_MeanAggregator` (#127550) is replaced by the vectorized implementation; master's `testGradient1D` is retained.
- The `third_party/xla/.../llvm/workspace.bzl` detour is gone; the now-redundant `np.negative` test edits (landed via #127561) are dropped.
- Diff is exactly `BUILD`, `nn_grad.py`, `nn_test.py`.

Validation against released TF 2.21.0 kernels with the rebased `nn_grad.py`:
- Reproducer: rank-1 `[4., 0.7, 2.]`, decreasing=True → gradient of `y[1]` is `[0., 0.5, 0.5]`; `(50,)` gradient shape is `(50,)`; rank-3 gradient shape matches the input; unknown-shape `def_function` no longer raises. Stock kernels 0/4, rebased 4/4.
- Isotonic test filter: stock 42 failed / 28 passed; rebased 30 passed / 56 subtests (`test_pooled_gradient`, `test_empty_gradient`, `test_gradient_unknown_shape`, `testGradient1D`).
- Bazel/GPU/TPU source-built checks are not available locally; GitHub CI covers linux cpu/cuda/cuda13, windows and arm64 build.

AI assistance: the original patch and this rebase/conflict resolution were implemented and tested with AI assistance (OpenAI Codex originally; rebase with opencode).

### CODE REVIEW

**Review by dmiltr3 — APPROVED** [review]
LGTM! Internal tests passed.

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thank you for your pull request!

Internal presubmit logic testing has successfully passed across all platforms. However, there is a minor Python linting error in `nn_test.py` that needs to be addressed before we can merge this PR.

**Issue**: The nested function `gradient(x)` inside `test_gradient_unknown_shape` triggers an `invalid-name` PyLint error because it does not conform to the test class naming conventions. 

**Expected Fix**: Please rename the `gradient` function inside `test_gradient_unknown_shape` to something like `ComputeGradient` or `_gradient` to satisfy the linter.

Here is an example patch of what needs to be changed in `tensorflow/python/ops/nn_test.py`:

```diff
@@ -2094,7 +2094,7 @@
   def test_gradient_unknown_shape(self):
     @def_function.function(
         input_signature=[tensor_spec.TensorSpec(None, dtypes.float64)])
-    def gradient(x):
+    def ComputeGradient(x):
       with backprop.GradientTape() as tape:
         tape.watch(x)
         y, _ = nn_ops.isotonic_regression(x)
         selected = y[..., 1]
       return tape.gradient(selected, x)
 
     for shape in [(3,), (2, 3), (2, 2, 3)]:
       x = constant_op.constant(np.broadcast_to([4., 0.7, 2.], shape))
       expected = np.broadcast_to([0., 0.5, 0.5], shape)
-      self.assertAllClose(expected, self.evaluate(gradient(x)))
+      self.assertAllClose(expected, self.evaluate(ComputeGradient(x)))
```

Once this is updated, the GitHub CI and our internal linters should pass, and I will be able to approve this!

**Comment by twelfthlabor** [comment]
Thx for the comment, the issue is addressed by using different naming convention.

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Hi @twelfthlabor, thank you for the updates!

Unfortunately, it looks like this PR now includes modifications to `third_party/xla/third_party/llvm/workspace.bzl`. 

The `third_party/xla/` directory is maintained in a separate upstream repository (`openxla/xla`) and is automatically vendored into TensorFlow via automated synchronization tooling. Changes made directly to this directory in the `tensorflow/tensorflow` repository will cause our PR sync to fail. 

Could you please revert the changes to `third_party/xla/third_party/llvm/workspace.bzl`? Once those changes are dropped, we should be able to proceed with merging this PR.

Thank you!

**Comment by twelfthlabor** [comment]
thx @dmiltr3 , changes for workspace.bzl are reverted, ready for another look.

**Review by dmiltr3 — DISMISSED** [review]
LGTM, thank you for the fix!

**Comment by twelfthlabor** [comment]
Rebased onto `master@72c0c0c9` and resolved the overlap with #127550: the vectorized `_MeanAggregator` now replaces the merged 1-D special case (covers 1-D, rank ≥ 3, empty and dynamic shapes; `testGradient1D` retained). The `workspace.bzl` detour and the duplicate `np.negative` hunks (#127561) are dropped. Verified against TF 2.21.0 kernels: repro 0/4 → 4/4, isotonic test filter 42 failed/28 passed → 30 passed/56 subtests. CI is re-running on the new head. Could you take another look when convenient, @dmiltr3? Thanks!

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thank you for the update!

The new test tests work well, but our internal presubmit checks identified one minor missing dependency due to the explicit usage of `tensorflow.python.eager.backprop` in the new tests. 

Could you please add `"//tensorflow/python/eager:backprop"` to the `deps` list for `nn_test` in `tensorflow/python/ops/BUILD`?

```diff
--- a/tensorflow/python/ops/BUILD
+++ b/tensorflow/python/ops/BUILD
@@ -3786,6 +3786,7 @@ cuda_py_strict_test(
         ":stateful_random_ops",
         ":variable_scope",
         ":variables",
+        "//tensorflow/python/eager:backprop",
         "//tensorflow/python/eager:context",
         "//tensorflow/python/eager:def_function",
         "//tensorflow/python/framework:constant_op",
```

Additionally, I noticed a pre-existing crash at HEAD in `testIncoorectKSize` which is causing one of the test shards to fail. This is unrelated to your PR, so I am ignoring it.

Once the `BUILD` file dependency is added, this is good to go!

**Comment by twelfthlabor** [comment]
I thought it's already in the current PR. Line 3789, can u double check pls?
https://github.com/twelfthlabor/tensorflow/blob/f8481721ea7df4c9cba569cc0ebc1a1a120e5f62/tensorflow/python/ops/BUILD#L3789

**Comment by twelfthlabor** [comment]
I see that the original 1-D Gradient bug is fixed [127550](https://github.com/tensorflow/tensorflow/pull/127550), current PR still addresses additional cases. 
 - Higher-rank inputs, such as 3-D tensors.
  - Empty inputs.
  - Shapes that aren’t known until runtime.
Lmk know if this PR should be closed, or good to proceed.

**Review by dmiltr3 — APPROVED** [review]
LGTM! Internal tests passed. Thank you for the contribution.

**Comment by nithyak0204** [comment]
Hi @twelfthlabor, Seems auto-merge is not happening but the changes are merged into master via commit [303300a](https://github.com/tensorflow/tensorflow/commit/303300abe50998de86b2acb09b3783f0502f3ed1), so we can close this PR.

Thank you for the contribution and helping improve TensorFlow!.
